source/websocket_decoder.c

```diff
@@ -0,0 +1,331 @@
+
+    memcpy(decoder->current_frame.masking_key + decoder->state_bytes_processed, data->ptr, bytes_to_consume);
+
+    aws_byte_cursor_advance(data, bytes_to_consume);
+    decoder->state_bytes_processed += bytes_to_consume;
+
+    /* If all bytes consumed, proceed to next state */
+    if (decoder->state_bytes_processed == 4) {
+        decoder->state++;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* PAYLOAD_BEGIN: Determine if we need to decode a payload. Consumes no data. */
+static int s_state_payload_begin(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data) {
+    (void)data;
+
+    /* Invoke on_frame() callback to inform user of non-payload data. */
+    int err = decoder->on_frame(&decoder->current_frame, decoder->user_data);
+    if (err) {
+        return AWS_OP_ERR;
+    }
+
+    /* Choose next state: either we have payload to process or we don't. */
+    if (decoder->current_frame.payload_length > 0) {
+        decoder->state_bytes_processed = 0;
+        decoder->state++;
+    } else {
+        decoder->state += 2;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* PAYLOAD_CONTINUE: Decode payload until we're done. */
+static int s_state_payload_continue(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data) {
+    if (data->len == 0) {
+        return AWS_OP_SUCCESS;
+    }
+
+    assert(decoder->current_frame.payload_length > decoder->state_bytes_processed);
+    uint64_t bytes_remaining = decoder->current_frame.payload_length - decoder->state_bytes_processed;
+    size_t bytes_to_consume = bytes_remaining < data->len ? (size_t)bytes_remaining : data->len;
+
+    struct aws_byte_cursor payload = aws_byte_cursor_advance(data, bytes_to_consume);
+
+    /* Unmask data, if necessary.
+     * RFC-6455 Section 5.3 Client-to-Server Masking
+     * Each byte of payload is XOR against a byte of the masking-key */
+    if (decoder->current_frame.masked) {
+        uint64_t mask_index = decoder->state_bytes_processed;
+
+        /* Optimization idea: don't do this 1 byte at a time */
+        uint8_t *current_byte = payload.ptr;
+        uint8_t *end_byte = payload.ptr + payload.len;
+        while (current_byte != end_byte) {
+            *current_byte++ ^= decoder->current_frame.masking_key[mask_index++ % 4];
+        }
+    }
+
+    /* TODO: validate utf-8 */
+
+    /* Invoke on_payload() callback to inform user of payload data */
+    int err = decoder->on_payload(payload, decoder->user_data);
+    if (err) {
+        return AWS_OP_ERR;
+    }
+
+    decoder->state_bytes_processed += payload.len;
+    assert(decoder->state_bytes_processed <= decoder->current_frame.payload_length);
+
+    /* If all data consumed, proceed to next state. */
+    if (decoder->state_bytes_processed == decoder->current_frame.payload_length) {
+        decoder->state++;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static state_fn *s_state_vtable[AWS_WEBSOCKET_DECODER_STATE_DONE] = {
```

Comment:
this isn't really a vtable, but i'm not good at naming things, so.... you decide.

Reply:
s_state_functions